Jazzz/readme revisions (#2)
xmtp/xmtp-memo-js: README.md

# XMTP-MEMO-JS

![Status](https://img.shields.io/badge/Project_Status-Developer_Preview-red)

## XMTP client SDK for pre-registration messaging

`xmtp-memo-js` provides a TypeScript implementation of an XMTP memo client.

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Author: Much clearer! 🙏

Reviewer: Aw great, @jazzz! The hard work is "writing down the bones," as they say -- helping to refine is a pleasure.

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Reviewer: Per feedback from Rich, called out the baseline requirement that an address must be registered to send and receive messages.

Per feedback from Yash, made an adjustment to the text to make the distinction between memos and messages early on - with your great "What is a memo?" section following up with the deets!

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Reviewer: Thank you for this, @jazzz! I think we are so close to landing on a precise pattern here - through your patience and diligence! I'm so excited about this possibility!

I love knowing that it is the address and not the account that is registered. 

On a related note - to help us arrive at a pattern - might we say that the account is sort of a "container concept" that includes the account address and account keys, for example?

If this is an acceptable concept =), how does this hypothesis land with you? Sometimes, we can refer more generically to the account, but when more precision is required to be accurate, we should refer to the address or keys. 

For example, see this doc that I worked with Martin to update: https://xmtp.org/docs/dev-concepts/key-generation-and-usage. Based on Martin's input, we applied the hypothesis above when we wrote the doc.

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Author: These are very helpful links! @jhaaaa  :pray:. 

I think embracing the use of the "Blockchain account" over "Wallet" has been helpful for our documentation. My goal is continue this by ensuring that I'm using clear and consistent language in further docs and features. (I've updated the question above to add some clarity)

> how does this hypothesis land with you? Sometimes, we can refer more generically to the account, but when more precision is required to be accurate, we should refer to the address or keys.  

On the surface it feels right. Using your hypothesis I might suggest that  "blockchain account must be registered" is a better fit than the "blockchain account address must be registered." 

I propose that 'registration' is the act of proving possession of a Blockchain Account (aka: privateKey) and verifying the intention to receive XMTP messages. The address in this case is not used directly in the process but is more of an implementation detail. 

Thoughts?

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Reviewer: Ah yes, I see the added `?`! 

And I'm so relieved =))) that "blockchain account must be registered" is accurate and what you prefer. This aligns with the patterns in the Key generation and usage doc. I think we're on the same page here - both in terms of technical accuracy and style. 

And thank you for your definition of what we mean by "registration." LOVE IT. 

Let me make the updates - and then I promise to step away from the vehicle and we can get this version out!

And separately, I'll update the XMTP content style guide to reflect this updated info.

> With XMTP, a blockchain account address must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered address to start communicating with an unregistered address by sending a memo instead of a message.

Reviewer: ```suggestion
With XMTP, a blockchain account must be registered on the XMTP network to send and receive end-to-end encrypted messages. Build your app with `xmtp-memo-js` to enable a registered account to start communicating with an unregistered account by sending a memo instead of a message.
```

:warning: :warning: :warning: `XMTP:Memos` are available as an early Developer Preview. :warning: :warning: :warning: <br>
The memo API will change without warning and may not be backward compatible.

### What is a memo?

> Memos are not the same as XMTP messages, though they are similar. For example, the payload structure of memos and messages is the same, but memos and messages are used in different circumstances.

Reviewer: I adjusted the text here to try to clarify that memos and messages do not carry the same payloads - literally. Like the exact content in a memo is not also carried in a message. I tried to do this by saying "payload structure" - but wanted ensure that my change didn't change the original meaning and create an inaccurate statement!

Reviewer: @jazzz - ah, I think I get it- thank you! What do you think of this edit below? And may I ask - what is the "Encoded Content" specification? Does this have to do with XMTP content types - or is this a more generic spec?

```suggestion
Memos are not the same as XMTP messages, though they are similar. For example, the payload structure of memos and messages is the same and they both use the "Encoded Content" specification. However, memos and messages are used in different circumstances.
```

Author: > What do you think of this edit below? 

I think I like the wording as is, actually. After the API goes through some feedback/testing we can return to make stronger claims about the interop between messages and memos. Currently in its untested state I'm hesitant to have aspirational docs.

By encoded content specification, I was referring to the way that data is encoded inside XTMP messages. It was an imprecise use of language as I think part of it is covered by the ContentTypes, but may include more.

Reviewer: Ah, got it... I took it too far! LOL I'll delete it now. <3

> - Memos allow a registered address to send messages that are **not end-to-end encrypted** to **any Ethereum Virtual Machine-compatible blockchain address**.

Reviewer: ```suggestion
- Memos allow a registered account to send messages that are **not end-to-end encrypted** to **any Ethereum Virtual Machine-compatible blockchain account**.
```

> - Messages allow a registered address to send **end-to-end encrypted messages** to **any address registered on the XMTP network**.

Reviewer: ```suggestion
- Messages allow a registered account to send **end-to-end encrypted messages** to **any account registered on the XMTP network**.
```

When possible, developers should always send messages with [XMTP-JS](https://github.com/xmtp/xmtp-js) to benefit from higher security and stability.

### Memo sequence

> 

Reviewer: Made tiny updates to the mermaid just to put the comment about Bola closer to Bola =) and as a comment about Amal's intention to message Bola. Also added names to the memo content to reinforce the sender and recipient.

```mermaid
sequenceDiagram
autonumber
    participant A as Amal
    participant X as XMTP
    participant B as Bola
    A->>X: <Register>
    Note left of A : Amal wants to message Bola
    Note right of B : 2: Bola isn't on the network. <br> Can't message.
    X -->> A: canMessage(Bola) : False
    Note left of  A : 3, 4: Send memo instead
    A->>X: Memo1: Hi Bola! It's Amal.
    A->>X: Memo2: Msg me when you get this
     Note over A,B: 1 month later
    B->>X: <Register>

    Note right of B : 6,7: Bola checks for memos
    X -->>B: Memo1: Hi Bola! It's Amal.
    X -->>B: Memo2: Msg me when you get this
     Note right of B : 8: Send E2E encrypted message to Amal
    B ->> X: Msg: Yo what's up Amal?
    X -->> A: Msg: Yo what's up Amal?
```

## :warning: :warning: :warning: Security notice :warning: :warning: :warning:

`xmtp-memo-js` uses [Lit protocol](https://developer.litprotocol.com/) to secure memos. Memos do not carry the same security characteristics as XMTP messages.

- Memos are **NOT** end-to-end encrypted. The memo encryption key is stored in the Lit Network and is only as secure as the escrow service.
- Memos have **NOT** undergone a formal security audit.

> ## Install

Reviewer: For the heading tweaks I made here and below, if you are interested, see the first row of the table here: https://developers.google.com/style/headings#heading-and-title-text

For doc sections that are about tasks - things someone needs to do - we use the "imperative mood." The form intends to make the title "action-oriented" - like a call-to-action on a UI button, for example - to help the user understand that they need to do something.

In the past, documentation used the gerund form (-ing) for task headings, but more recently the style has changed to move away from them. Tech writing trends! ;-)

    participant X as XMTP
    participant B as Bola
    A->>X: <Register>
    Note left of A : Amal wants to message Bola

>     Note right of B : 2: Bola isn't on the network. <br> Can't message.

Author: Thats a nice touch!

```

## Send a memo

> Memos don't require that the recipient address be registered on the XMTP network. Here is an example Memo payload:

Reviewer: ```suggestion
Memos don't require that the recipient account be registered on the XMTP network. Here is an example Memo payload:
```

```ts

> // Send a memo. There is no requirement that this address be registered on the XTMP network.

Reviewer: ```suggestion
// Send a memo. There is no requirement that this account be registered on the XTMP network.
```

memoClient.sendMemo('0x0000000000000000000000000000000000000000', "Gm!")
```

## List memos

> Here is an example of how to retrieve and list memos for an address:

Reviewer: ```suggestion
Here is an example of how to retrieve and list memos for an account:
```
